src/main/java/spoon/support/visitor/equals/EqualsVisitor.java

```diff
@@ -96,9 +100,12 @@ public void biScan(CtRole role, CtElement element, CtElement other) {
 		if (element == other) {
 			return;
 		}
-
 		try {
 			lastRole = role;
+			if (haveDifferentDeclaringInnerClasses(element, other)) {
+				fail(role, element, other);
+				return;
+			}
 			super.biScan(element, other);
```

Comment:
Traversing the whole hierarchy up on every element seems a bit redundant and could probably affect performance quite a bit.

Reply:
Thanks, I moved the new check out from the recursive call